test/parallel_api/ranges/std_ranges_test.h

```diff
@@ -0,0 +1,455 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#include <oneapi/dpl/execution>
+
+#include "support/test_config.h"
+
+#include "support/utils.h"
+
+#if _ENABLE_STD_RANGES_TESTING
+
+#include <oneapi/dpl/ranges>
+#include <span>
+#include <iostream>
+#include <vector>
+#include <typeinfo>
+#include <string>
+
+namespace test_std_ranges
+{;
+
+#if _ONEDPL_HETERO_BACKEND
+template<int call_id = 0>
+auto dpcpp_policy()
+{
+    auto exec = TestUtils::default_dpcpp_policy;
+    using Policy = decltype(exec);
+    return TestUtils::make_new_policy<TestUtils::new_kernel_name<Policy, call_id>>(TestUtils::default_dpcpp_policy);
+}
+#endif //_ONEDPL_HETERO_BACKEND
+
+auto host_policies() { return std::true_type{};}
+
+enum TestDataMode
+{
+    data_in,
+    data_in_out,
+    data_in_in,
+    data_in_in_out
+};
+
+auto f_mutuable = [](auto&& val) { return val *= val; };
+auto proj_mutuable = [](auto&& val) { return val *= 2; };
+
+auto f = [](auto&& val) { return val * val; };
+auto proj = [](auto&& val){ return val * 2; };
+auto pred = [](auto&& val) { return val == 5; };
+auto pred_2 = [](auto&& val1, auto&& val2) { return val1 == val2; };
+
+auto pred1 = [](auto&& val) -> decltype(auto) { return val > 0; };
+auto pred2 = [](auto&& val) -> decltype(auto) { return val == 4; };
+auto pred3 = [](auto&& val) -> decltype(auto) { return val < 0; };
+
+struct P2
+{
+    P2(auto v): x(v) {}
+    int x = {};
+    int y = {};
+
+    int proj() { return x; }
+    friend bool operator==(const P2& a, const P2& b) { return a.x == b.x && a.y == b.y; }
+};
+
+template<typename DataType, typename Container, TestDataMode Ranges = data_in, bool RetTypeCheck = true>
+struct test
+{
+    template<typename Policy, typename Algo, typename... Args>
+    std::enable_if_t<std::is_same_v<Policy, std::true_type>>
+    operator()(Policy, Algo algo, Args... args)
+    {
+        operator()(oneapi::dpl::execution::seq, algo, args...);
+        operator()(oneapi::dpl::execution::unseq, algo, args...);
+        operator()(oneapi::dpl::execution::par, algo, args...);
+        operator()(oneapi::dpl::execution::par_unseq, algo, args...);
+    }
+
+    template<typename Policy, typename Algo, typename Checker, typename Transform>
+    std::enable_if_t<!std::is_same_v<Policy, std::true_type> && Ranges == data_in>
+    operator()(Policy&& exec, Algo algo, Checker checker, Transform tr, auto... args)
+    {
+        constexpr int max_n = 10;
```

Comment:
Testing only one very small size is not a sufficient for a non-experimental feature. Could you extend the testing 
 with more sizes?

Reply:
I think in given cases it is enough, because the adding algorithms are not new algorithms, it is just new high level API that re-calls the existing implementation, which already has test coverage with non-trivial data sizes in particular.

Reply:
I am still inclined towards having more coverage.
Ranges bring more ways to pass the data into backend implementations. You can also pass a projection in contrast to the iterator-based counterparts. I think it deserves testing coverage at least like a single type in iterator-based tests.
I am worried that some implementations, for example, multiple-work-group ones for device policies, remain uncovered and it may hide issues specific to those unique traits of the ranges.

Reply:
> Ranges bring more ways to pass the data into backend implementations

Please give me an example when test coverage is not sufficient? I mean an example when iterator based algo version is passed, range based version (which is introduced by this PR) is passed, but there is still an error in the impl(?)

In regard to  `multiple-work-group ones for device policies` and other differentiations by data sizes. A projection is  applied within lambda/functor wrapper which is introduced on frontend level (high level API), so it is immediately diagnosed with any data sizes.
Perhaps just there is one exception - algo `sort`. Here we have projection applying within backend.. But that case also is covered by existed `sort_by_key` test.

The second reason why I would not like introducing big data sizes for ranges API - it consume additional infrastructure resources and testing time.

Reply:
> Please give me an example when test coverage is not sufficient

It is indeed hard to find such an issue;  I cannot do it. Perhaps, if some compiler optimization breaks the correctness for a case with a non-trivial projection, but it is too far-fetched.
From time consumption point, I suppose that adding one or two more sizes is negligible.

So, I agree to leave it as is. At least, If somebody detects a non-covered issue regarding sizes (which is unlikely), then the tests can be extended.